src/libproc_macro/bridge/server.rs

```diff
@@ -0,0 +1,242 @@
+        }
+
+        $(impl<S: $name> $name for MarkedTypes<S> {
+            $(fn $method(&mut self, $($arg: $arg_ty),*) $(-> $ret_ty)* {
+                <_>::mark($name::$method(&mut self.0, $($arg.unmark()),*))
+            })*
+        })*
+    }
+}
+with_api!(Self, self_, define_mark_types_impls);
+
+struct Dispatcher<S: Types> {
+    handle_store: HandleStore<S>,
+    server: S,
+}
+
+macro_rules! define_dispatcher_impl {
+    ($($name:ident {
+        $(fn $method:ident($($arg:ident: $arg_ty:ty),* $(,)*) $(-> $ret_ty:ty)*;)*
+    }),* $(,)*) => {
+        trait DispatcherTrait {
+            // HACK(eddyb) these are here to allow `Self::$name` to work below.
+            $(type $name;)*
+            fn dispatch(&mut self, b: Buffer<u8>) -> Buffer<u8>;
+        }
+
+        impl<S: Server> DispatcherTrait for Dispatcher<MarkedTypes<S>> {
+            $(type $name = <MarkedTypes<S> as Types>::$name;)*
+            fn dispatch(&mut self, mut b: Buffer<u8>) -> Buffer<u8> {
+                let Dispatcher { handle_store, server } = self;
+                // FIXME(eddyb) shouldn't be needed post-NLL.
+                let (tag, tag_len) = {
+                    let mut b2 = &b[..];
+                    (api_tags::Method::decode(&mut b2, &mut ()), b.len() - b2.len())
+                };
+                match tag {
+                    $(api_tags::Method::$name(m) => match m {
+                        $(api_tags::$name::$method => {
+                            let r = panic::catch_unwind(panic::AssertUnwindSafe(|| {
+                                let mut _b = &b[tag_len..];
+                                reverse_decode!(_b, handle_store; $($arg: $arg_ty),*);
+                                $name::$method(server, $($arg),*)
+                            })).map_err(PanicMessage::from);
+                            b.clear();
+                            r.encode(&mut b, handle_store);
+                        })*
+                    }),*
+                }
+                b
+            }
+        }
+    }
+}
+with_api!(Self, self_, define_dispatcher_impl);
+
+impl Bridge {
+    fn run_server<
+        S: Server,
+        I: Encode<HandleStore<MarkedTypes<S>>>,
+        O: for<'a, 's> DecodeMut<'a, 's, HandleStore<MarkedTypes<S>>>,
+    >(
+        handle_counters: &'static client::HandleCounters,
+        server: S,
+        input: I,
+        f: impl FnOnce(Bridge, Buffer<u8>) -> Buffer<u8> + Send + 'static,
+    ) -> Result<O, PanicMessage> {
+        use std::sync::mpsc::{channel, Receiver, Sender};
+
+        let mut dispatcher = Dispatcher {
+            handle_store: HandleStore::new(handle_counters),
+            server: MarkedTypes(server),
+        };
+
+        let mut b = Buffer::new();
+        input.encode(&mut b, &mut dispatcher.handle_store);
+
+        let (req_tx, req_rx) = channel();
+        let (res_tx, res_rx) = channel();
+
+        let join_handle = thread::spawn(|| {
```

Comment:
I wasn't able to quite determine why a `thread::spawn` was here, is there a reason that this is needed? It seems great to me that it works like this because it's a great proof that this works, but for efficiency I'd imagine that we probably want to avoid this.

Reply:
I wanted a crater run with thread isolation, just to see what would break. I don't want people to rely on TLS in proc macros, basically. With a bit more work IPC could be used too.

I think that if we don't want to do this thread spawning thing, we should have it as an option that the server can set, so we can keep both codepaths.

The good news is that nothing outside of this function needs to change if we want cross-thread vs single thread.